src/compiler/crystal/zero_one_or_many.cr

```diff
@@ -0,0 +1,110 @@
+# An Array(T)-like type that's optimized for the case of
+# frequently having zero or one elements.
+#
+# It's a struct, so if you have a ZeroOneOrMany and you
+# want to mutate, you have to use `with` or `without`
+# and overwrite the value you have with what you get
+# from those methods.
+#
+# Warning: you should never keep the original value
+# around when using the above methods as the internal
+# array might be mutated (for performance reasons),
+# so the original value might mutate in that case!
+#
+# TODO: add tests for this type.
+struct ZeroOneOrMany(T)
```

Comment:
I don't think this should be a top-level type.

Reply:
Thanks!